qcodes/dataset/measurements.py

```diff
@@ -669,7 +677,35 @@ def register_parameter(
 
         return self
 
-    def _register_parameter(self : T, name: str,
+    @staticmethod
+    def _infer_paramtype(parameter: _BaseParameter,
+                         paramtype: Optional[str]) -> str:
+        """
+        Infer the best parameter type to store the parameter supplied.
+
+        Args:
+            parameter: The parameter to to infer the type for
+            paramtype: The initial supplied parameter type or None
+
+        Returns:
+            The inferred parameter type. If a parameter type is supplied this
+            will overwrite the inferred parameter type.
+        """
+        if isinstance(parameter.vals, vals.Arrays) and paramtype is None:
+            paramtype = 'array'
+        elif isinstance(parameter, ArrayParameter) and paramtype is None:
+            paramtype = 'array'
+        elif isinstance(parameter.vals, vals.Strings) and paramtype is None:
+            paramtype = 'text'
+        # TODO complex case here once support is added
+        # should we try to figure out if parts of a multiparamter are
+        # arrays or something else?
+        # default to numeric as before
+        elif paramtype is None:
```

Comment:
I suggest to move the `if paramtype is None` to the caller - then the code of the inference method is purely about inferring.